### ISSUE TITLE
Omniauth error

### ISSUE THREAD

**Issue by joshreeves** [Shopify/shopify_app#230]
I've set bundler to use the latest code from the repository, but now when I attempt to install or view the app I see a flash of a page with a single HTML link `Continue` which displays for a few seconds and then the entire browser refreshes and displays this Rails error: `OmniAuth::Strategies::OAuth2::CallbackError`.

I did not have this problem when using the latest gem. It seems to install properly and I have sometimes seen a flash of content coming from the API, but the error seems to always happen. Any help would be greatly appreciated.

Here is my `initializers/omniauth.rb` file:
```
Rails.application.config.middleware.use OmniAuth::Builder do
  provider :shopify,
    ShopifyApp.configuration.api_key,
    ShopifyApp.configuration.secret,
    redirect_uri: ShopifyApp.configuration.redirect_uri,
    scope: ShopifyApp.configuration.scope
end
```
And my `initializers/shopify_app.rb`:
```
ShopifyApp.configure do |config|
  config.api_key = ENV['SHOPIFY_CLIENT_API_KEY']
  config.secret = ENV['SHOPIFY_CLIENT_API_SECRET']
  config.redirect_uri = "[domain]/auth/shopify/callback"
  config.scope = "read_content, write_content, read_themes, write_themes, read_products, write_products, read_orders, write_orders, read_customers, write_customers, read_script_tags, write_script_tags"
  config.embedded_app = true
  config.webhooks = [
    {topic: 'products/delete', address: 'http://requestb.in/17y2ze91', format: 'json'},
  ]
end
```

**Comment by kevinhughes27** [comment]
What is your system and browser info and is this an embedded app? @nwtn could be related to your changes.

@joshreeves we're going to be QAing the new version before shipping next week. I'd recommend using a released version until then.

**Comment by nwtn** [comment]
Yup sounds like it's related to my change. What app is it? If possible I'd love to test it with the problematic app itself. +1 to using a released version.

**Comment by joshreeves** [comment]
Embedded app in Latest Chrome on Mac OSX. It's an unpublished development app. If you need any more info let me know. I'll likely switch back to the gem. Thanks for the quick response!

**Comment by kevinhughes27** [comment]
hmm okay I am also using latest chrome on osx.

I just ran through a tests creating and installing a new app using - localhost, forward and ngrok. All worked fine for me. I thought your scope might be an issue since there was some recent changes there but I tried with your scope and couldn't reproduce.

Can you create an example app that reproduces the problem and include a screenshot of the setup of your api client in the partners area (redact and keys). Then I can look into this more.

**Comment by joshreeves** [comment]
Strange. It's working for me this morning. Can't figure out what is different but I'll keep an eye on it and let you know if it happens again.

**Comment by joshreeves** [comment]
Hi Kevin. Just updated to the latest published gem and the Omniauth issue is back for me. I created a barebones app following the steps in the quickstart guide, and was able to reproduce the bug. I'll provide all of the steps and as much additional details as possible:

1. Create barebones app following quickstart guide (here is the one I created: https://github.com/joshreeves/shopify-test)
2. Set up that app settings page as follows:
![image](https://cloud.githubusercontent.com/assets/2991100/13902224/b575f14c-edfd-11e5-9325-e694187b5f97.png)
3. Using Chrome 49.0.2623.87 (64-bit) in Mac OSX 10.10.5 Yosemite, install the app into a Shopify store. (At this point, it might appear that everything is working).
4. Exit out of the Chrome application completely.
5. Log back into the store admin, click Apps and then click on the test app. You should see the word "Continue" appear temporarily and then one of a few different Omniauth errors. The behavior at this point seems fairly erratic and seems to occasionally result in a successful authentication if you hit the back button and re-attempt accessing the app a few times.

Hope this helps.

**Comment by kevinhughes27** [comment]
Thanks @joshreeves I'll checkout out your repo in a couple of days (I am at Unite 2016 this week). 
Thanks for doing that - it really helps me in the debugging process.

**Comment by theducksnuts** [comment]
Hi guys, I'm also experiencing this "Continue" link in my App - I have a theory on how it's caused that you may be able to try replicate - as it doesn't always seem to happen. I am constantly logging in and out of different stores while testing/building - it seems as though this Continue link only occurs when there may be some conflict of this. Although I must add - my continue link doesn't then result in an error - it stalls it a few seconds, then loads the page in my App. Also using latest Gem in Chrome but on W10. Interested in your thoughts on this as although it's not causing errors, I would like it removed before releasing my App. Any other Q's of info let me know.

**Comment by kevinhughes27** [comment]
The continue link is not an errors its part of the redirect page that escapes the esdk. I agree though it looks kind of weird. @nwtn does the link really serve a purpose? perhaps we should just remove it.

**Comment by nwtn** [comment]
The idea behind the link was to provide a fallback if both `meta` and JS redirects failed. That’s a pretty specific edge case, so I’m happy to remove the link and only add it back if this shows up as a legitimate problem.

**Comment by kevinhughes27** [comment]
Yeah I think it might be too specific - the context in which this executes is inside the Shopify admin which has modern browser and js requirements so we should be fine.

**Comment by nwtn** [comment]
Not to go on too much of a tangent, but the idea was in cases of broken JS moreso than JS being turned off; we’ve had a few cases of JS breakages recently, and I put this in for resiliency/robustness in situations like that. But anyway, I digress; it probably is overkill for here at the moment.

**Comment by joshreeves** [comment]
FWIW, when the JS failed in the admin panel a week or so ago, my embedded apps didn't work at all. The existence of the `Continue` button didn't mitigate that. Sorry I can't provide any additional useful details.

**Comment by kevinhughes27** [comment]
Yeah lets remove it. @nwtn can you make the change?

**Comment by nwtn** [comment]
maybe later this week, but if it’s urgent it might be easier/faster for somebody else to grab this

**Comment by kevinhughes27** [comment]
no rush!

**Comment by forsbergplustwo** [comment]
Hey Guys,

At the same time this issue was created, the Shopify Order Printer app began having some new bugs related to the login phase:

1- When selecting an Order in Shopify Admin, and then choosing 'Print with Order Printer', the app will start login, show the 'Continue' link, then instead of showing the selected order/orders, just show the latest order.

2- When selecting an Order in Shopify Admin, and then choosing 'Print with Order Printer' it would begin login but then redirect to the app store page for the app.

I've been in contact with support and got a reply from Keith that Point 2 had been fixed, however the fix seems to be... just do the same as Point 1 ( ignore selected order and just show the latest)

I bring this up here because the app worked perfectly before, but since the changes to Shopify App gem at the same time as this issue was logged.

I have a lot of experience eith the Shopify Order Printer app as an app developer (5 years with Shopify apps), but obviously don't have access to it's code.

I created both the Order Printer Templates app (creates compatible code for Order Printer) and the free Order Printer Emailer app (takes any Order Printer compatible liquid code and generates them as PDF, so they can be created from emails etc).. so I have reverse engineered the internals of the Shopify app and have the resources/knowledge to help.

Would be great if either:

a) You guys with great experience with the login process take a look at the Shopify Order Printer app

Or

b) Open source the Order Printer code, like the PixelPrinter app that it replaced back in the day was.

I'd be more than happy to look into this issue, as well as a couple of other bugs that are in the app. Most likely easy fixed but judt not getting a priority ;)

I say this because Order Printer is getting a lot of negative reviews lately, which all mention the issues above.. just look at anything beloew 5 star here: https://apps.shopify.com/order-printer#reviews-heading

Many thanks,
Bjorn

**Comment by kevinhughes27** [comment]
@EiNSTeiN- since you worked on Order Printer last - seems like there are some issues with it. I don't know if they are actually related to ShopifyApp or not though.

**Comment by EiNSTeiN-** [comment]
I believe the Order Printer issues are unrelated to the original issue reported here.

**Comment by uurcankaya** [comment]
@kevinhughes27 has continue link removed?

**Comment by kevinhughes27** [comment]
ping @nwtn

**Comment by nwtn** [comment]
@uurcankaya should be gone in latest version, let me know if it's still showing up for you

**Comment by uurcankaya** [comment]
i am also confused with updating the app with the new gem, it used to create errors when I used generators with new gem, i remember I ended up creating the app from scratch How can I update an existing app? Is there an easy way @kevinhughes27 @nwtn

**Comment by nwtn** [comment]
@uurcankaya You should be able to upgrade by updating the version for the `gem 'shopify_app'` line in your Gemfile and running `bundle install`.

**Comment by kevinhughes27** [comment]
yeah updating the gem is all you need to do. Occasionally the generators change but we try to keep all logic inside the gem itself so that upgrading is easy

**Comment by EiNSTeiN-** [comment]
I'm going to close this issue since I think the original issue was fixed by #243.
* For any new issue in existing Shopify apps, they can be reported through our support.
* For any new issue with this gem specifically, open a  new issue in github describing the problem.
